providers/implementations/digests/sha2_prov.c

```diff
@@ -56,41 +57,350 @@ static int sha1_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+    c->md_len = (unsigned int)val;
+
+    if (c->md_len != md_len) {
+        OPENSSL_cleanse(c, sizeof(c));
+        return 0;
+    }
+
+    return 1;
+}
+
+static const unsigned char sha512magic[] = "SHA512v1";
+#define SHA512_SERIALIZATION_LEN \
+    ( \
+     (sizeof(sha512magic) - 1) /* magic */ \
+     + sizeof(uint64_t) * 8 /* c->h */ \
+     + sizeof(uint64_t) * 2 /* c->Nl + c->Nh */ \
+     + SHA512_CBLOCK /* c->u.d/c->u.p */ \
+     + sizeof(uint64_t) * 2 /* c->num + c->md_len */ \
+     )
+
+static int SHA512_Serialize(SHA512_CTX *c, unsigned char *output,
+                            size_t *outlen)
+{
+    unsigned char *p;
+    unsigned long i;
+
+    if (output == NULL) {
+        if (outlen == NULL)
+            return 0;
+
+        *outlen = SHA512_SERIALIZATION_LEN;
+        return 1;
+    }
+
+    if (outlen != NULL && *outlen < SHA512_SERIALIZATION_LEN)
+        return 0;
+
+    p = output;
+
+    /* Magic code */
+    memcpy(p, sha512magic, sizeof(sha512magic) - 1);
+    p += sizeof(sha512magic) - 1;
+
+    /* h */
+    for (i = 0; i < sizeof(c->h) / sizeof(SHA_LONG64); i++)
+        p = OPENSSL_store_u64_le(p, c->h[i]);
+
+    /* Nl, Nh */
+    p = OPENSSL_store_u64_le(p, c->Nl);
+    p = OPENSSL_store_u64_le(p, c->Nh);
+
+    /* data */
+    memcpy(p, c->u.p, SHA512_CBLOCK);
+    p += SHA512_CBLOCK;
+
+    /* num, md_len */
+    p = OPENSSL_store_u64_le(p, c->num);
+    p = OPENSSL_store_u64_le(p, c->md_len);
+
+    if (outlen != NULL)
+        *outlen = SHA512_SERIALIZATION_LEN;
+
+    return 1;
+}
+
+static int SHA512_Deserialize(SHA512_CTX *c, const unsigned char *input,
+                              size_t len, unsigned int md_len)
+{
+    const unsigned char *p;
+    uint64_t val;
+    unsigned long i;
+
+    if (c == NULL || input == NULL || len != SHA512_SERIALIZATION_LEN)
+        return 0;
+
+    /* Magic code */
+    if (memcmp(input, sha512magic, sizeof(sha512magic) - 1) != 0)
+        return 0;
+
+    p = input + sizeof(sha512magic) - 1;
+
+    /* h */
+    for (i = 0; i < (sizeof(c->h) / sizeof(SHA_LONG64)); i++) {
+        p = OPENSSL_load_u64_le(&val, p);
+        c->h[i] = (SHA_LONG64)val;
+    }
+
+    /* Nl, Nh */
+    p = OPENSSL_load_u64_le(&val, p);
+    c->Nl = (SHA_LONG64)val;
+    p = OPENSSL_load_u64_le(&val, p);
+    c->Nh = (SHA_LONG64)val;
+
+    /* data */
+    memcpy(c->u.p, p, SHA512_CBLOCK);
+    p += SHA512_CBLOCK;
+
+    /* num, md_len */
+    p = OPENSSL_load_u64_le(&val, p);
+    c->num = (unsigned int)val;
+    p = OPENSSL_load_u64_le(&val, p);
+    c->md_len = (unsigned int)val;
+
+    if (c->md_len != md_len) {
+        OPENSSL_cleanse(c, sizeof(c));
+        return 0;
+    }
+
+    return 1;
+}
+
+static OSSL_FUNC_digest_settable_ctx_params_fn sha2_settable_ctx_params;
+static OSSL_FUNC_digest_gettable_ctx_params_fn sha2_gettable_ctx_params;
+
+static const OSSL_PARAM known_sha2_settable_ctx_params[] = {
+    {OSSL_DIGEST_SERIALIZATION, OSSL_PARAM_OCTET_STRING, NULL, 0, 0},
```

Comment:
Should serialisation/de-serialisatiob really be via a param?
I'd have thought they deserve provider API support and new EVP APIs to access them.